### PR TITLE
feat(vfs): add PRAGMA-based time travel for querying historical database states

### LINKED ISSUE #849: feat(vfs): Add custom SQLite functions for time travel queries
## Overview

Add custom SQLite functions to the Litestream VFS that enable time travel queries, allowing users to set a target timestamp and query historical database state at that point in time.

## Motivation

Users need the ability to interactively query their database at different points in time without restoring multiple copies. The VFS already has the infrastructure to read data at any timestamp (via `CalcRestorePlan`), but currently only uses it once at database open.

## Proposed SQL Interface

```sql
-- Set viewing time to a specific timestamp
SELECT litestream_set_time('2024-01-01T12:00:00Z');

-- Query data at that time
SELECT * FROM users WHERE id = 123;

-- Check what time we're currently viewing
SELECT litestream_current_time();

-- Return to the latest available time
SELECT litestream_reset_time();
```

### Key Insight
The VFS already has all necessary infrastructure:
- `CalcRestorePlan()` (replica.go:499-572) supports timestamp-based filtering of LTX files
- `buildIndex()` constructs page-to-LTX mappings
- `ReadAt()` fetches pages using the index

We need to expose runtime control of the target timestamp.

#### 1. C Extension Infrastructure (src/litestream-vfs.c)
- Use `sqlite3_auto_extension()` to register a callback for all new connections
- Register custom functions via `sqlite3_create_function()` in the callback
- Create C wrapper functions that call exported Go functions
- Pass VFSFile context through `sqlite3_user_data()`

Example C code:
```c
extern void GoLitestreamSetTime(void* vfsFile, const char* timestamp);

void litestream_set_time_impl(sqlite3_context* ctx, int argc, sqlite3_value** argv) {
    void* vfsFile = sqlite3_user_data(ctx);
    const char* timestamp = (const char*)sqlite3_value_text(argv[0]);
    GoLitestreamSetTime(vfsFile, timestamp);
}
```

#### 2. Go Implementation (vfs.go)

Add to `VFSFile` struct:
```go
targetTime *time.Time  // nil = latest time
```

Implement methods:
```go
// SetTargetTime rebuilds the index for a specific point in time
func (f *VFSFile) SetTargetTime(t time.Time) error {
    f.mu.Lock()
    defer f.mu.Unlock()
    
    // Call CalcRestorePlan with target time
    infos, err := CalcRestorePlan(ctx, f.vfs.ReplicaClient, 0, t, f.vfs.logger)
    if err != nil {
        return err
    }
    
    // Rebuild index from filtered LTX files
    if err := f.buildIndex(infos); err != nil {
        return err
    }
    
    // Clear cache - pages are now stale
    f.cache.Purge()
    
    f.targetTime = &t
    return nil
}

// GetTargetTime returns the current target time (nil = latest)
func (f *VFSFile) GetTargetTime() *time.Time {
    f.mu.Lock()
    defer f.mu.Unlock()
    return f.targetTime
}

// ResetTime returns to the latest available time
func (f *VFSFile) ResetTime() error {
    f.mu.Lock()
    defer f.mu.Unlock()
    f.targetTime = nil
    // Rebuild index with all available files
    return f.rebuildIndexToLatest()
}
```

#### 3. CGO Exports (cmd/litestream-vfs/main.go)

Export functions for C to call:
```go
//export GoLitestreamSetTime
func GoLitestreamSetTime(vfsFilePtr unsafe.Pointer, timestampCStr *C.char) C.int {
    timestamp := C.GoString(timestampCStr)
    t, err := time.Parse(time.RFC3339, timestamp)
    if err != nil {
        return -1
    }
    
    vfsFile := (*VFSFile)(vfsFilePtr)
    if err := vfsFile.SetTargetTime(t); err != nil {
        return -1
    }
    return 0
}

//export GoLitestreamCurrentTime
func GoLitestreamCurrentTime(vfsFilePtr unsafe.Pointer, outBuf *C.char, bufSize C.int) C.int {
    vfsFile := (*VFSFile)(vfsFilePtr)
    t := vfsFile.GetTargetTime()
    
    var result string
    if t == nil {
        result = "latest"
    } else {
        result = t.Format(time.RFC3339)
    }
    
    // Copy to output buffer
    // ...
    return 0
}

//export GoLitestreamResetTime
func GoLitestreamResetTime(vfsFilePtr unsafe.Pointer) C.int {
    vfsFile := (*VFSFile)(vfsFilePtr)
    if err := vfsFile.ResetTime(); err != nil {
        return -1
    }
    return 0
}
```

#### 4. Connection-to-VFSFile Mapping

Since SQLite functions receive a context but need to find the VFSFile:
- Maintain a global map: `connectionToFile sync.Map` mapping `sqlite3*` to `*VFSFile`
- Register mapping in `VFS.Open()` when file is opened
- Clean up in `VFSFile.Close()`
- Use `sqlite3_context_db_handle()` in C code to get connection handle

### Phase 1: Core Go Implementation
1. Add `targetTime` field to `VFSFile` struct
2. Implement `SetTargetTime()`, `GetTargetTime()`, `ResetTime()` methods
3. Add connection-to-file mapping infrastructure
4. Add unit tests for time-based index rebuilding

### Phase 2: C Extension Integration
1. Modify `src/litestream-vfs.c` to register auto-extension callback
2. Add C wrapper functions for the three SQL functions
3. Export Go functions with `//export` directive
4. Update Makefile if needed for CGO linking

### Phase 3: Testing
1. Integration tests with time travel scenarios
2. Test concurrent connections with different target times
3. Test edge cases:
   - Timestamp before first LTX file
   - Timestamp after last LTX file
   - Invalid timestamp format
   - Null/empty timestamp
4. Performance testing with large databases

### Phase 4: Documentation
1. Update VFS documentation with time travel examples
2. Add usage examples to README
3. Document limitations and performance characteristics

## Files to Modify

- `vfs.go` - Add time travel methods to VFSFile
- `src/litestream-vfs.c` - Register SQL functions
- `cmd/litestream-vfs/main.go` - Export Go functions for C
- `cmd/litestream-vfs/main_test.go` - Add integration tests
- `Makefile` - Update if CGO linking changes needed
- `docs/` - Add time travel documentation

## Key Constraints

1. **Read-Only**: Time travel is naturally read-only (VFS doesn't support writes)
2. **Connection-Isolated**: Each connection maintains its own target time
3. **Transaction-Granular**: Can only travel to times where LTX files exist (transaction boundaries)
4. **Performance**: Index rebuild is expensive - warn users about this
5. **Cache Invalidation**: All cached pages must be cleared on time change

## Alternative/Additional Approach

Consider also supporting query parameter-based time setting for simpler use cases:
```bash
sqlite3 "file:/path/to/db?vfs=litestream&time=2024-01-01T00:00:00Z"
```

This would be simpler to implement (no C code) but less flexible (can't change time mid-session).

## References

- SQLite create_function: https://sqlite.org/c3ref/create_function.html
- mattn/go-sqlite3 RegisterFunc: https://pkg.go.dev/github.com/mattn/go-sqlite3#SQLiteConn.RegisterFunc (not applicable here since we're building a loadable extension)
- Existing time-based restoration: `replica.go:499-572` (CalcRestorePlan)
- VFS implementation: `vfs.go`
- Extension entry point: `src/litestream-vfs.c:11-22`

## Testing Checklist

- [ ] Unit tests for SetTargetTime/GetTargetTime/ResetTime
- [ ] Integration test: set time, query data, verify historical state
- [ ] Test with multiple concurrent connections at different times
- [ ] Test timestamp before first transaction
- [ ] Test timestamp after last transaction
- [ ] Test invalid timestamp formats
- [ ] Test performance with large databases (>1GB)
- [ ] Test cache invalidation on time change
- [ ] Test lock state handling during index rebuild
- [ ] Verify works with standalone sqlite3 CLI
- [ ] Verify works with Python sqlite3 module

### PR DESCRIPTION
**Pull request by corylanou**
## Summary

Implements time travel queries for the Litestream VFS using SQLite PRAGMA commands, allowing users to view and query database state at specific points in time.

### PRAGMA Commands

```sql
-- Set viewing time to a specific timestamp
PRAGMA litestream_time = '2024-01-01T12:00:00Z';

-- Query data at that time
SELECT * FROM users WHERE id = 123;

-- Check current effective time
PRAGMA litestream_time;

-- Return to latest state
PRAGMA litestream_time = latest;
```

### Why PRAGMA Instead of SQL Functions?

The original implementation used custom SQL functions (`litestream_set_time()`, etc.), but this approach had drawbacks:
- SELECT statements that modify state are semantically incorrect
- Incompatible with read-only tools like Datasette that expect SELECT to be side-effect-free
- Required complex connection-to-file mapping infrastructure

PRAGMA commands are the standard SQLite mechanism for configuration and state changes, making this approach:
- **Semantically correct**: PRAGMAs are designed for state changes
- **Tool-compatible**: Works with Datasette and other read-only query tools
- **Cleaner architecture**: No connection mapping needed; FileControl receives file directly

### Architecture

The implementation uses SQLite's `xFileControl` mechanism with `SQLITE_FCNTL_PRAGMA`:

1. **sqlite3vfs fork**: Added `FileController` interface and CGO bridge for FileControl operations
2. **VFSFile.FileControl()**: Handles `litestream_time` PRAGMA with set/query/reset operations
3. **Index rebuilding**: Uses existing `CalcRestorePlan()` and `rebuildIndex()` infrastructure
4. **Cache invalidation**: Clears LRU cache when switching time views

### Files Modified

- **vfs.go**: Add `FileControl()` method implementing `FileController` interface
- **src/litestream-vfs.c**: Simplified to only register VFS (removed SQL functions)
- **go.mod**: Add replace directive for sqlite3vfs fork with FileControl support

### Files Removed (Simplified Architecture)

- **vfs_connections.go**: Connection mapping no longer needed
- **cmd/litestream-vfs/vfs_cgo.go**: File ID extraction no longer needed

## Dependencies

This PR depends on a fork of `psanford/sqlite3vfs` that adds FileControl support:
- Fork: `github.com/corylanou/sqlite3vfs` branch `vfs-time-travel-pragma`
- Changes: Added `FileController` interface and `goVFSFileControl` CGO bridge

## Documentation

Documentation for VFS time travel will be added to the litestream.io repository.
See [litestream.io#144](https://github.com/benbjohnson/litestream.io/issues/144) for tracking.

## Test Plan

- [x] Set time via PRAGMA and query historical data
- [x] Query current time setting via PRAGMA
- [x] Reset to latest state via PRAGMA
- [x] Verify unknown PRAGMAs pass through to SQLite normally

### Manual Testing

```bash
# Build VFS extension
CGO_ENABLED=1 go build -tags vfs -o bin/litestream-vfs ./cmd/litestream-vfs

# Test time travel
sqlite3 <<EOF
.load ./bin/litestream-vfs
.open 'file:test.db?vfs=litestream'
PRAGMA litestream_time = '2024-01-01T00:00:00Z';
SELECT * FROM users;
PRAGMA litestream_time;
PRAGMA litestream_time = latest;
EOF
```

## Limitations

1. **Read-Only**: Time travel is read-only (VFS doesn't support writes)
2. **Transaction Granular**: Can only travel to times where LTX files exist
3. **Performance**: Index rebuild is expensive for large databases

Closes #849

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: vfs.go [hunk 1]
```diff
@@ -194,16 +199,77 @@ func (f *VFSFile) Open() error {
...
+	}
+
+	var pos ltx.Pos
+	if len(infos) > 0 {
+		pos = ltx.Pos{TXID: infos[len(infos)-1].MaxTXID}
+	}
+
+	maxTXID1 := maxLevelTXID(infos, 1)
+
+	f.mu.Lock()
```
Comment (benbjohnson) [inline vfs.go:244]: Doing a `defer f.mu.Unlock()` after this seems safer than doing the unlock at the end.

**Review by benbjohnson — APPROVED** [review]
